### PR TITLE
tls: throw an error on getLegacyCipher

### PR DESCRIPTION
**Pull request by jasnell**
throw an error on unknown getLegacyCipher input
rather than returning the default. Makes the
behavior more explicit and matches up with the
behavior on the command line switches.

### CODE REVIEW

**Comment by jasnell** [comment]
@misterdjules @mdawsonibm

**Comment by misterdjules** [comment]
Can we add a test to make sure that `getLecagyCiphers` throws/doesn't throw as expected?

**Comment by jasnell** [comment]
Done

**Comment by jasnell** [comment]
https://github.com/joyent/node/issues/15445 @misterdjules

**Comment by mdawsonibm** [comment]
One minor nit.  Should the tests validate the type of the Error, ie a TypeError or an Error as appropriate ? Otherwise lgtm.

**Comment by mdawsonibm** [comment]
One minor nit.  Should the tests validate the type of the Error, ie a TypeError or an Error as appropriate ? Otherwise lgtm.

**Comment by jasnell** [comment]
@mhdawson @misterdjules ok, made the additional edits discussed and squashed the commits. Please verify.

File: lib/_tls_wrap.js [hunk 1]
```diff
@@ -836,10 +836,13 @@ exports.connect = function(/* [port, host], options, cb */) {
 
   var defaults = {
     rejectUnauthorized: '0' !== process.env.NODE_TLS_REJECT_UNAUTHORIZED,
-    ciphers: tls.DEFAULT_CIPHERS,
     checkServerIdentity: tls.checkServerIdentity
   };
 
+  if (tls.DEFAULT_CIPHERS != tls.getLegacyCiphers('v0.10.38')) {
```
Comment (misterdjules) [inline lib/_tls_wrap.js]: If one passes `--cipher-list='ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:RC4:HIGH:!MD5:!aNULL:!EDH'`, my understanding is that `tls.DEFAULT_CIPHERS` will === `tls.getLegacyCiphers('v0.10.38'))`, and `default.ciphers` won't be set. I haven't tested it, so it may be a wrong assumption.

So I guess that here what is meant is more `if (user didn't pass --enable-legacy-cipher-list=v0.10.38) { ... }`, which is not equivalent to `if (tls.DEFAULT_CIPHERS != tls.getLegacyCiphers('v0.10.38'))`.

Thoughts?

Comment (misterdjules) [inline lib/_tls_wrap.js]: Also, in general please use strict equality unless not possible. In this occurence, please use `!==` instead of `!=`.

Comment (jasnell) [inline lib/_tls_wrap.js]: Yes, that is the case. I personally don't see this as a significant problem. If someone passes in the exact default cipher list as v0.10.38 then the likely intent is to revert the behavior back to v0.10.38, which means the default.ciphers not being set. It's an acceptable edge case IMHO but I'm good either way. It's a bit of a larger change tho since this does not yet specifically track whether `--enable-legacy-cipher-list` is passed vs. `--cipher-list`

File: test/simple/test-tls-cipher-list.js [hunk 2]
```diff
@@ -40,6 +40,54 @@ function doTest(checklist, env, useswitch) {
       // Test NODE_LEGACY_CIPHER_LIST
       if (env) options = {env:{"NODE_LEGACY_CIPHER_LIST": env}};
       break;
+    case 4:
```
Comment (misterdjules) [inline test/simple/test-tls-cipher-list.js]: Having arbitrary numbers determine the behavior of tests seems very confusing to me, because it's not clear when I read `doTest(V1038Ciphers, 'v0.10.38', 8);` that we test that `NODE_LEGACY_CIPHER_LIST takes precendence over NODE_CIPHER_LIST`.

Can we refactor that to make it clearer? Maybe we could have separate function clearly named for each test we want to perform? Parameters for each test would be local to these functions and make it clear what each test does.

File: doc/api/tls.markdown [hunk 3]
```diff
@@ -223,10 +252,15 @@ Example:
 
 ## tls.getLegacyCiphers(version)
 
-Returns the legacy default cipher suite for the specified Node.js release.
+Returns a default cipher list used in a previous version of Node.js. The
+version parameter must be a string whose value identifies previous Node.js
+release version. The only values currently supported are `v0.10.38`, `v0.10.39`, `v0.12.2` and `v0.12.3`.
 
-Example:
-    var cipher_suite = tls.getLegacyCiphers('v0.10.38');
```
Comment (misterdjules) [inline doc/api/tls.markdown]: Keeping an example (with code formatting) would be nice.

**Comment by jasnell** [comment]
Closing this as it cannot land here now. Will be refactoring this change and opening a new PR against master on nodejs/node.
